### PR TITLE
Make errors on listener non-fatal

### PR DESCRIPTION
**Pull request by tomaka**
In #1218, we decided to make listener errors non-fatal.
There was however something missing, and errors were actually fatal.
This fixes it.

### CODE REVIEW

**Comment by tomaka** [comment]
> effectively this means that listeners must never fail.

I think we debated this in the past at the time of #1218, and the conclusion was that yes, listeners must never fail. There isn't really a fundamental reason why a listener would fail (as an example, the TCP one doesn't), and in these rare situations I think it's fine to log an error and either return `None` or return `Poll::Pending` all the time.

**Comment by twittner** [comment]
If listeners must not fail, would this not better be expressed by a `Stream<Item = ListenerEvent<Self::ListenerUpgrade>>` or, if they should be able to report errors, by a `Stream<Item = ListenerEvent<Self::ListenerUpgrade, Self::Error>>`?

**Comment by tomaka** [comment]
I was almost finished with the changes when I realized a problem:
`TcpListener::bind` is asynchronous, and if it fails we return the error as part of the listener.
In this specific situation, it should indeed be fatal.

**Comment by twittner** [comment]
> Should we go all the way and make `listen_on` return a `Future` to a `Stream`?

Do you mean
```rust
Future<Output = Result<Stream<Item = ListenerEvent<Self::ListenerUpgrade, Self::Error>>, Self::Error>
```
?

**Comment by tomaka** [comment]
Yes. It seems like a lot of complexity and would require adding yet another associated type to `Transport`.

**Comment by twittner** [comment]
I am in favour of this proposal.

**Comment by tomaka** [comment]
I reworked this PR to instead add a `ListenerEvent::Error` variant that is non-fatal.

File: core/src/nodes/listeners.rs [hunk 1]
```diff
@@ -269,15 +269,17 @@ where
...
+                    let id = *listener_project.id;
+                    self.listeners.push_front(listener);
+                    return Poll::Ready(ListenersEvent::Error {
+                        listener_id: id,
+                        error,
                     })
                 }
-                Poll::Ready(Some(Err(err))) => {
-                    return Poll::Ready(ListenersEvent::Error {
+                Poll::Ready(None) | Poll::Ready(Some(Err(_))) => {
```
Comment (twittner) [inline core/src/nodes/listeners.rs]: I would be nice if we could log this error should it occur.

File: core/src/transport.rs [hunk 2]
```diff
@@ -318,6 +343,27 @@ impl<T> ListenerEvent<T> {
             None
         }
     }
+
+    /// Returns `true` if this is an `Error` listener event.
+    pub fn is_non_fatal_error(&self) -> bool {
```
Comment (twittner) [inline core/src/transport.rs]:
```suggestion
    pub fn is_error(&self) -> bool {
```

File: core/src/transport.rs [hunk 3]
```diff
@@ -318,6 +343,27 @@ impl<T> ListenerEvent<T> {
...
+        } else {
+            false
+        }
+    }
+
+    /// Try to turn this listener event into the `Error` part.
+    ///
+    /// Returns `None` if the event is not actually a `Error`,
+    /// otherwise the error.
+    pub fn into_non_fatal_error(self) -> Option<TErr> {
```
Comment (twittner) [inline core/src/transport.rs]:
```suggestion
    pub fn into_error(self) -> Option<TErr> {
```

File: core/src/transport/dummy.rs [hunk 4]
```diff
@@ -56,7 +56,7 @@ impl<TOut> Clone for DummyTransport<TOut> {
 impl<TOut> Transport for DummyTransport<TOut> {
     type Output = TOut;
     type Error = io::Error;
-    type Listener = futures::stream::Pending<Result<ListenerEvent<Self::ListenerUpgrade>, io::Error>>;
+    type Listener = futures::stream::Pending<Result<ListenerEvent<Self::ListenerUpgrade, io::Error>, io::Error>>;
```
Comment (twittner) [inline core/src/transport/dummy.rs]:
```suggestion
    type Listener = futures::stream::Pending<Result<ListenerEvent<Self::ListenerUpgrade, Self::Error>, Self::Error>>;
```

File: transports/tcp/src/lib.rs [hunk 5]
```diff
@@ -93,7 +93,7 @@ impl $tcp_config {
 impl Transport for $tcp_config {
     type Output = $tcp_trans_stream;
     type Error = io::Error;
-    type Listener = Pin<Box<dyn Stream<Item = Result<ListenerEvent<Self::ListenerUpgrade>, io::Error>> + Send>>;
+    type Listener = Pin<Box<dyn Stream<Item = Result<ListenerEvent<Self::ListenerUpgrade, Self::Error>, io::Error>> + Send>>;
```
Comment (twittner) [inline transports/tcp/src/lib.rs]:
```suggestion
    type Listener = Pin<Box<dyn Stream<Item = Result<ListenerEvent<Self::ListenerUpgrade, Self::Error>, Self::Error>> + Send>>;
```

**Pull request merged**